### PR TITLE
Fix conversation ID validation to return 400 instead of 500 for long IDs

### LINKED ISSUE #10489: [Bug]: [API] GET /api/conversations/{id} with long id returns 500 (expected 4xx JSON)
### Is there an existing issue for the same bug? (If one exists, thumbs up or comment on the issue instead).

- [x] I have checked the existing issues.

### Describe the bug and reproduction steps

# 🐞 [Bug]: [API] GET /api/conversations/{id} with long id returns 500 (expected 4xx JSON)
---
### Checklist
- [x] I searched existing issues and discussions.
- [x] I can reproduce on the latest local build.
- [x] Not related to model/provider credentials or quotas.

## 📜 Description
Requests with a very long {id} to GET /api/conversations/{id} return 500 instead of a 4xx JSON error.  
Expected: reject oversized IDs early and return 400/414 with a JSON error body.

---

## 📋 Environment & Setup

| Component            | Detail                                                                 |
|---------------------|-------------------------------------------------------------------------|
| 🚀 Deployment        | Docker (CLI)                                                            |
| 🔐 Auth              | Not required locally                                                    |
| 💻 macOS Version     | macOS Sequoia 15.4.1                                                    |
| 🐍 Python Version    | 3.12.x                                                                  |
| 🐳 Docker Version    | Docker Desktop 4.43.2 (199162)                                          |
| 📦 OpenHands Image   | `docker.all-hands.dev/all-hands-ai/openhands:0.53`                      |
| 📎 OpenHands Digest  | `docker.all-hands.dev/all-hands-ai/openhands@sha256:61b05c5d247d12f7b490d945f63216e25f7cbda1a5232671a09cc7bf3226ce92` |
| 🔧 Runtime Image     | `docker.all-hands.dev/all-hands-ai/runtime:0.53-nikolaik`               |
| 🤖 Model             | `openrouter/deepseek/deepseek-r1-0528:free`                             |

---

### Endpoint
`GET /api/conversations/{id}`

### STEP 1
```bash
# Generate a 5,000-character 'a' string
LONG=$(python3 -c "print('a'*5000)")
```
### STEP 2
```bash
# Hit the /conversations/ endpoint
curl -i "${API_BASE_URL:-http://localhost:3000}/api/conversations/$LONG"
```

---

## 🧯 Observed Behavior

- `500 Internal Server Error`
```
curl -i "${API_BASE_URL:-http://localhost:3000}/api/conversations/$LONG"
HTTP/1.1 500 Internal Server Error
date: Mon, 18 Aug 2025 22:34:17 GMT
server: uvicorn
content-length: 21
content-type: text/plain; charset=utf-8

Internal Server Error
```

---

## ✅ Expected Behavior
- A safe **4xx** response (e.g., `414 URI Too Long` or `400 Bad Request`).
- **JSON** error body only (no HTML).

---

### OpenHands Installation

Docker command in README

### OpenHands Version

0.53

### Model Name

openrouter/deepseek/deepseek-r1-0528:free

### Operating System

MacOS

### Logs, Errors, Screenshots, and Additional Context

_No response_

### PR DESCRIPTION
**Pull request by amanape**
**DESCRIPTION GENERATED BY OPENHANDS**

- [x] This change is worth documenting at https://docs.all-hands.dev/
- [x] Include this change in the Release Notes. If checked, you **must** provide an **end-user friendly** description for your change below

**End-user friendly description of the problem this fixes or functionality this introduces.**

Fixed an issue where API requests with very long or malformed conversation IDs would return a `500 Internal Server Error` instead of a proper `400 Bad Request` error. The API now properly validates conversation IDs and returns clear error messages for invalid requests, improving the developer experience and API reliability.

---
**Summarize what the PR does, explaining any non-trivial design decisions.**

This PR adds comprehensive validation for conversation IDs across all API endpoints to prevent 500 errors and improve security. The implementation uses a clean, simple approach with FastAPI's dependency system.

**Key Changes:**
1. **Added `validate_conversation_id()` function** in `openhands/server/utils.py` that validates:
   - Length (max 100 characters to prevent resource exhaustion)
   - Path traversal attempts (blocks `../`, `/`, `\` characters)
   - Control characters (blocks newlines, tabs, null bytes)
   - Returns proper `400 Bad Request` HTTPException for invalid IDs

2. **Updated all conversation endpoints** to use automatic validation:
   ```python
   @app.get('/conversations/{conversation_id}')
   async def get_conversation(
       conversation_id: str = Depends(validate_conversation_id),  # ← Simple & clean
       conversation_store: ConversationStore = Depends(get_conversation_store),
   ) -> ConversationInfo | None:
       # conversation_id is a regular str, validation already happened
   ```

3. **Applied to all conversation endpoints**:
   - `GET /api/conversations/{conversation_id}`
   - `DELETE /api/conversations/{conversation_id}`
   - `GET /api/conversations/{conversation_id}/remember-prompt`
   - `POST /api/conversations/{conversation_id}/start`
   - `POST /api/conversations/{conversation_id}/stop`
   - `PATCH /api/conversations/{conversation_id}`
   - `POST /api/conversations/{conversation_id}/exp-config`

4. **Added comprehensive test suite** covering edge cases and security scenarios

**Design Decisions:**
- **Simple Dependency Approach**: Uses `str = Depends(validate_conversation_id)` for clean, readable code
- **No Custom Types**: Avoided unnecessary complexity - functions work with regular `str` parameters
- **DRY Principle**: Single validation function, no repeated code in route handlers
- **100-character limit**: Balances flexibility (normal UUIDs are 32-36 chars) with security
- **Security-first approach**: Blocks common attack vectors proactively

**Before vs After:**
```python
# Before: Manual validation in every handler (repeated 7+ times)
@app.get('/conversations/{conversation_id}')
async def get_conversation(conversation_id: str, ...):
    conversation_id = validate_conversation_id(conversation_id)  # ❌ Repeated
    # ... business logic

# After: Automatic validation via dependency (0 repetition)
@app.get('/conversations/{conversation_id}')
async def get_conversation(
    conversation_id: str = Depends(validate_conversation_id),  # ✅ Automatic
    ...
):
    # ... business logic (no manual validation needed)
```

**Security Improvements:**
- Prevents path traversal attacks (`../../../etc/passwd`)
- Blocks control character injection (newlines, tabs, null bytes)
- Prevents resource exhaustion from extremely long IDs
- Maintains backward compatibility for all valid conversation IDs

**Technical Benefits:**
- **Automatic Application**: All endpoints get validation without manual calls
- **FastAPI Integration**: Proper dependency injection and error handling
- **Error Consistency**: Same 400 error format across all endpoints
- **Maintainability**: Changes to validation logic only need to be made in one place
- **Type Safety**: Functions work with clean `str` types
- **Performance**: Validation happens once at the dependency level

---
**Link of any specific issues this addresses:**

Fixes #10489

---

To run this PR locally, use the following command:

GUI with Docker:
```
docker run -it --rm   -p 3000:3000   -v /var/run/docker.sock:/var/run/docker.sock   --add-host host.docker.internal:host-gateway   -e SANDBOX_RUNTIME_CONTAINER_IMAGE=docker.all-hands.dev/all-hands-ai/runtime:15704cd-nikolaik   --name openhands-app-15704cd   docker.all-hands.dev/all-hands-ai/openhands:15704cd
```

CLI with uvx:
```
uvx --python 3.12 --from git+https://github.com/All-Hands-AI/OpenHands@fix-conversation-id-validation-issue-10489 openhands
```

### CODE REVIEW

File: openhands/server/utils.py [hunk 1]
```diff
@@ -13,6 +14,63 @@
...
+            status_code=status.HTTP_400_BAD_REQUEST,
+            detail='Conversation ID contains control characters',
+        )
+
+    return conversation_id
+
+
+# Custom type annotation for validated conversation IDs
+# This combines FastAPI's Path parameter with automatic validation
+ValidatedConversationId = Annotated[
```
Comment (enyst) [inline openhands/server/utils.py]: Maybe we don’t need a custom type?

I wonder, as long as the Depends was on the endpoint, it would be nice to continue working with a str if we can

Comment (amanape) [reply openhands/server/utils.py]: Agreed

**Review by enyst — APPROVED** [review]
Thank you!
